text/0000-io-safety.md

```diff
@@ -0,0 +1,429 @@
+
+Memory safety is elevated above other programming hazards because it isn't
+just about avoiding unintended behavior, but about avoiding situations where
+it's impossible to bound the set of things that a piece of code might do.
+
+I/O safety is also in this category, for two reasons.
+
+ - I/O safety errors can lead to memory safety errors in the presence of
+   safe wrappers around `mmap` (on platforms with OS-specific APIs allowing
+   them to otherwise be safe).
+
+ - I/O safety errors can also mean that a piece of code can read, write, or
+   delete data used by other parts of the program, without naming them or
+   being given a reference to them. It becomes impossible to bound the set
+   of things a crate can do without knowing the implementation details of all
+   other crates linked into the program.
+
+Raw handles are much like raw pointers into a separate address space; they can
+dangle or be computed in bogus ways. I/O safety is similar to memory safety;
+both prevent spooky-action-at-a-distance, and in both, ownership is the main
+foundation for robust abstractions, so it's natural to use similar safety
+concepts.
+
+[`std::mem::forget` being safe]: https://doc.rust-lang.org/std/mem/fn.forget.html
+[changed to safe]: https://rust-lang.github.io/rfcs/1066-safe-mem-forget.html
+
+## I/O Handles as plain data
+
+The main alternative would be to say that raw handles are plain data, with no
+concept of I/O safety and no inherent relationship to OS resource lifetimes. On
+Unix-like platforms at least, this wouldn't ever lead to memory unsafety or
+undefined behavior.
+
+However, most Rust code doesn't interact with raw handles directly. This is a
+good thing, independently of this RFC, because resources ultimately do have
+lifetimes, so most Rust code will always be better off using higher-level types
+which manage these lifetimes automatically and which provide better ergonomics
+in many other respects. As such, the plain-data approach would at best make raw
+handles marginally more ergonomic for relatively uncommon use cases. This would
+be a small benefit, and may even be a downside, if it ends up encouraging people
+to write code that works with raw handles when they don't need to.
+
+The plain-data approach also wouldn't need any code changes in any crates. The
+I/O safety approach will require changes to Rust code in crates such as
+[`socket2`], [`nix`], and [`mio`] which have APIs involving [`AsRawFd`] and
+[`RawFd`], though the changes can be made gradually across the ecosystem rather
+than all at once.
+
+## New types for `RawFd`/`RawHandle`/`RawSocket`
+
+Some comments on [rust-lang/rust#76969] suggest introducing new wrappers
+around the raw handles. Completely closing the safety loophole would also
+require designing new traits, since `AsRaw*` doesn't have a way to limit the
+lifetime of its return value. This RFC doesn't rule this out, but it would be a
+bigger change.
```

Comment:
It would, but I personally think it would be a preferable one. I would love to see AsFd, returning an Fd with a lifetime borrowed from the object, as well as an "into" version of the method that drops the object and returns an owned Fd (lifetime 'static). Other code that wanted to be safe could accept Fd.

I feel like the trait proposed here is too narrow for such a broad name.

Reply:
I agree, I think it would be good to use the established Rust type mechanisms for this. But I am also very biased when it comes to types and I don't actually have to work with the resulting set of new types. ;) See [here](https://internals.rust-lang.org/t/pre-rfc-i-o-safety/14585/18?u=ralfjung) for the slightly longer version of this.

Regarding the RFC document, I think it would be good to include this snippet of code here in this "alternatives" section, just so that it is unambiguously clear what this proposal would look like:
```rust
/// Owning an instance of this type means owning the right to perform I/O on the underlying FD
/// for the duration of `'fd`.
struct BorrowedFd<'fd> {
    fd: RawFd,
    _lifetime: PhantomData<&'fd RawFd>, // Lifetime of the file descriptor we're borrowing from
}

trait AsBorrowedFd {
    fn as_borrow_fd<'fd>(&'fd self) -> BorrowedFd<'fd>;
}
```

Also I think it is worth noting that `AsRawFd+IoSafe` is basically the same trait as `AsBorrowedFd`, just without type safety being enforced by the compiler.

Reply:
A `BorrowedFd` would also have the advantage that you could turn it into a `BorrowedWrapper<T> where T: FromRawFd` that internally uses `ManuallyDrop`, e.g. `UdpSocket` -> `BorrowedFd<'a>` -> `BorrowedWrapper<'a, File>`.

This has uses when temporarily converting from one fileish type into another so you can get its safe rust wrapper methods that operate on the descriptor instead of going through libc, e.g. when you want to get the metadata for a descriptor.

Reply:
I don't understand what `BorrowedWrapper` would look like, or what it would be good for.

`BorrowedFd<'a>` is just the direct translation of `&'a T` into the I/O safety world.

Reply:
The equivalent to `BorrowedWrapper<U>` would be `Deref<Target = U>`, perhaps I got the generics wrong here. The idea is to have go from a struct that owns a file descriptor to a borrowed file descriptor and then to a temporarily constructed struct of a different type that doesn't run its drop (i.e. doesn't close the file descriptor).

Example use: 

https://github.com/rust-lang/rust/blob/f77b1a5973521bce26d92d346304cfaca05ae79e/library/std/src/sys/unix/kernel_copy.rs#L470-L477

Reply:
And `BorrowedFd` may not really be necessary. Fileish structs generally carry the `RawFd` inside, so we can just return `&RawFd` directly. That's what `AsRawFd` should have done all along.

Reply:
`&RawFd` doesn't suffice, because RawFd is just a u32, and you can get a u32 from a `&u32`.

Reply:
@the8472 makes a good point though; `&OwnedFd` could work? Then we wouldn't need separate `OwnedFd` and `BorrowedFd` types. (We'd still have separate traits of course.)

Reply:
It looks like the representation does matter, as it's really nice to be able to use `repr(transparent)` with these types, since that means we can pass them directly to FFI.

@the8472 I agree that a `Deref<Target = U>` returning a temporary non-dropping view is useful though, and I think we can do it as a layer on top of `BorrowedFd` et al ([unsafe-io does something similar)](https://docs.rs/unsafe-io/0.6.9/unsafe_io/trait.AsUnsafeFile.html#method.as_file_view). I've now added a basic implementation of this to io-experiment, with an example [here](https://github.com/sunfishcode/io-experiment/blob/main/examples/portable-views.rs#L12).